src/kmer.jl

```diff
@@ -0,0 +1,346 @@
+
+"Shortcut for the type `RNAKmer{27,1}`"
+const RNAKmer27 = RNAKmer{27,1}
+
+"Shortcut for the type `RNAKmer{31,1}`"
+const RNAKmer31 = RNAKmer{31,1}
+
+"Shortcut for the type `RNAKmer{63,2}`"
+const RNAKmer63 = RNAKmer{63,2}
+
+"Shortcut for the type `Kmer{AminoAcidAlphabet,K,N}`"
+const AAKmer{K,N} = Kmer{AminoAcidAlphabet,K,N}
+
+const DNACodon = DNAKmer{3,1}
+const RNACodon = RNAKmer{3,1}
+
+###
+### Base Functions
+###
+
+@inline ksize(::Type{Kmer{A,K,N}}) where {A,K,N} = K
+@inline capacity(::Type{Kmer{A,K,N}}) where {A,K,N} = div(64N, bits_per_symbol(A()))
+@inline capacity(seq::Kmer) = capacity(typeof(seq))
+@inline n_unused(::Type{Kmer{A,K,N}}) where {A,K,N} = capacity(Kmer{A,K,N}) - K
+@inline n_unused(seq::Kmer) = n_unused(typeof(seq))
+
+"""
+    checkmer(::Type{Kmer{A,K,N}}) where {A,K,N}
+
+Internal method - enforces good kmer type parameterisation.
+
+For a given Kmer{A,K,N} of length K, the number of words used to
+represent it (N) should be the minimum needed to contain all K symbols,
+no larger (wasteful) no smaller (just... wrong).
+
+Because it is used on type parameters / variables, these conditions should be
+checked at compile time, and the branches / error throws eliminated when the
+parameterisation of the Kmer type is good. 
+"""
+@inline function checkmer(::Type{Kmer{A,K,N}}) where {A,K,N}
+    if K < 1
+        throw(ArgumentError("Bad kmer parameterisation. K must be greater than 0."))
+    end
+    n = seq_data_len(A, K)
+    if n !== N
+        # This has been significantly changed conceptually from before. Now we
+        # don't just check K, but *enforce* the most appropriate N for K.
+        throw(ArgumentError("Bad kmer parameterisation. For K = $K, N should be $n"))
+    end
+end
+
+@inline Base.length(x::Kmer{A,K,N}) where {A,K,N} = K
+@inline Base.summary(x::Kmer{A,K,N}) where {A,K,N} = string(eltype(x), ' ', K, "-mer")
+
+function Base.typemin(::Type{Kmer{A,K,N}}) where {A,K,N}
+    return Kmer{A,K,N}(ntuple(i -> zero(UInt64), N))
+end
+
+function Base.typemax(::Type{Kmer{A,K,N}}) where {A,K,N}
+    return Kmer{A,K,N}((typemax(UInt64), ntuple(i -> typemax(UInt64), N - 1)...))
+end
+
+function Base.rand(::Type{Kmer{A,K,N}}) where {A,K,N}
+    return Kmer{A,K,N}((rand(UInt64), ntuple(i -> rand(UInt64), N - 1)...))
+end
+
+Base.rand(::Type{Kmer{A,K}}) where {A,K} = rand(kmertype(Kmer{A,K}))
+
+function Base.rand(::Type{T}, size::Integer) where {T<:Kmer}
+    return [rand(T) for _ in 1:size]
+end
+
+###
+### Old Mer Base Functions - not transferred to new type.
+###
+#@inline encoded_data_type(::Type{Mer{A,K}}) where {A,K} = UInt64
+#@inline encoded_data_type(::Type{BigMer{A,K}}) where {A,K} = UInt128
+#@inline encoded_data_type(x::AbstractMer) = encoded_data_type(typeof(x))
+#@inline encoded_data(x::AbstractMer) = reinterpret(encoded_data_type(typeof(x)), x)
+#@inline ksize(::Type{T}) where {A,K,T<:AbstractMer{A,K}} = K
+#@inline Base.unsigned(x::AbstractMer) = encoded_data(x)
```

Comment:
Some commented out sections here, and below, and also a bunch of whitespace above this.